DS -> FSDP and back again (#2133)
huggingface/blog: deepspeed-to-fsdp-and-back.md

## Are FSDP and DeepSpeed Interchangeable?

Recently we tried running a training pipeline with DeepSpeed and PyTorch FSDP. We noticed that the results obtained differed. The specific model was Mistral-7B base and it was loaded in half-precision (`bfloat16`). While the DeepSpeed (blue) loss had converged well, the FSDP (orange) loss was not decreasing, as can be seen in Figure 1.

[[ZACH: INSERT FIGURE TWO HERE]]

We hypothesized that the learning rate may need scaling by the number of GPUs and bumped up the learning rate by 4x since we were using 4 GPUs. Then, we saw the following loss behavior, shown in Figure 2. 

> It looked like the desired behavior had been achieved by scaling the FSDP learning rate by the number of GPUs! However,  when we tried a different learning rate (`1e-5`) without scaling, we observed similar loss and gradient norm characteristics for both frameworks, shown in Figure 3.

Reviewer: ```suggestion
It looked like the desired behavior had been achieved by scaling the FSDP learning rate by the number of GPUs! However, when we tried a different learning rate (`1e-5`) without scaling, we observed similar loss and gradient norm characteristics for both frameworks, shown in Figure 3.
```

| Training Step (forward, backward, reduction)                                             | ❌         | FSDP<br>DeepSpeed | follows [fsdp.MixedPrecision](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.MixedPrecision)<br>follows `deepspeed_config_file` mixed precision settings |
| Optimizer (pre-step)                                                                     | ✅         | FSDP<br>DeepSpeed | upcasting (if any) to `torch_dtype`<br>upcasting everything to `float32`                                                                                                   |
| Optimizer (actual step)                                                                  | ✅         | FSDP<br>DeepSpeed | occurs in `torch_dtype`<br>occurs in `float32`                                                                                                                             |

> Table 1: Summary of how FSDP and DeepSpeed handle mixed precision

A few takeaway points:

> * As noted in the 🤗  Documentation, a rule of thumb when performing mixed precision is to keep trainable parameters in torch.float32. 

Reviewer: ```suggestion
* As noted in the 🤗 Accelerate documentation, a rule of thumb when performing mixed precision is to keep trainable parameters in torch.float32. 
```

* Upcasting, as is done in `DeepSpeed`, may have negligible effect on memory consumption when sharding over a large number of GPUs. However, when using `DeepSpeed` on a small number of GPUs, the 2x increase in memory consumption can be significant.

> * The torch-native implementation of FSDP does not force upcasting, allowing a user to operate PyTorch Optimizers  in low precision. This offers more flexibility than the native upcasting of `DeepSpeed`.

Reviewer: ```suggestion
* The torch-native implementation of FSDP does not force upcasting, allowing a user to operate PyTorch optimizers in low precision. This offers more flexibility than the native upcasting of `DeepSpeed`.
```

## Harmonizing DeepSpeed and FSDP in 🤗 Accelerate

> To better align DeepSpeed and FSDP in 🤗 Accelerate, we can perform upcasting automatically for FSDP when mixed precision is enabled. We created a pull request with this change that was included  in the [0.30.0 release](https://github.com/huggingface/accelerate/releases/tag/v0.30.0).

Reviewer: ```suggestion
To better align DeepSpeed and FSDP in 🤗 Accelerate, we can perform upcasting automatically for FSDP when mixed precision is enabled. We created a pull request with this change that was included in the [0.30.0 release](https://github.com/huggingface/accelerate/releases/tag/v0.30.0).
```

[[ZACH: INSERT FIGURE FOUR HERE]]

The result of this PR is to allow FSDP to operate in two modes:

> -  A “mixed-precision” mode like the DeepSpeed counterpart

Reviewer: ```suggestion
- A “mixed-precision” mode like the DeepSpeed counterpart
```

- A low precision mode for memory constrained scenarios, as shown in Figure 5.

The two new FSDP modes are summarized in Figure 6 and compared with DeepSpeed.
| **Framework**             | **Model Loading (`torch_dtype`)** | **Mixed Precision** | **Preparation (Local)** | **Training** | **Optimizer (Local)** |
| ------------------------- | --------------------------------- | ------------------- | ----------------------- | ------------ | --------------------- |
| FSDP (memory-constrained) | `bf16`                            | default (none)      | `bf16`                  | `bf16`       | `bf16`                |

> | FSDP (aligned mode)       | `bf16`                            | `bf16`              | `fp32`                  | `bf16`       | `fp32`                |

Reviewer: Is the "aligned mode" the same as the "mixed-precision" mode? I think it'd be clearer to use the same language if they are

> ---

Reviewer 2: Need to update https://github.com/huggingface/blog/blob/main/_blog.yml as well

> ---

Reviewer 2: This is missing a thumbnail (which should be hosted in this same repo)

title: "From DeepSpeed to FSDP and Back Again with Hugging Face Accelerate"
authors:
- user: mirinflim
- user: aldopareja
- user: muellerzr

> - user: stas00

Reviewer 2: For people outside HF, please also add `guest: true` and specify their org as in https://github.com/huggingface/blog/blob/main/assisted-generation-support-gaudi.md?plain=1

---

# A Hugging Face Accelerate Story of Multiple Backends: FSDP and DeepSpeed

> There are two popular implementations of the [ZeRO Redundancy Optimizer (Zero)](https://arxiv.org/abs/1910.02054) algorithm in the community, one from [DeepSpeed](https://github.com/microsoft/DeepSpeed) and the other from [PyTorch](https://pytorch.org/docs/stable/fsdp.html). Hugging Face [Accelerate](https://huggingface.co/docs/accelerate/en/index) exposes both these frameworks for the end users to train/tune their models. This blog highlights the differences between how these backends are exposed through Accelerate. To enable users to seamlessly switch between these backends, we [upstreamed a precision related change](https://github.com/huggingface/accelerate/issues/2624) and a [concept guide](https://huggingface.co/docs/accelerate/concept_guides/fsdp_and_deepspeed).

Reviewer 2: ```suggestion
There are two popular implementations of the [ZeRO Redundancy Optimizer (Zero)](https://arxiv.org/abs/1910.02054) algorithm in the community, one from [DeepSpeed](https://github.com/microsoft/DeepSpeed) and the other from [PyTorch](https://pytorch.org/docs/stable/fsdp.html). Hugging Face [Accelerate](https://huggingface.co/docs/accelerate/en/index) exposes both these frameworks for the end users to train/tune their models. This blog highlights the differences between how these backends are exposed through Accelerate. To enable users to seamlessly switch between these backends, we [upstreamed a precision-related change](https://github.com/huggingface/accelerate/issues/2624) and a [concept guide](https://huggingface.co/docs/accelerate/concept_guides/fsdp_and_deepspeed).
```

## Are FSDP and DeepSpeed Interchangeable?

> Recently we tried running a training pipeline with DeepSpeed and PyTorch FSDP. We noticed that the results obtained differed. The specific model was Mistral-7B base and it was loaded in half-precision (`bfloat16`). While the DeepSpeed (blue) loss had converged well, the FSDP (orange) loss was not decreasing, as can be seen in Figure 1.

Reviewer 2: ```suggestion
Recently, we tried running a training pipeline with DeepSpeed and PyTorch FSDP. We noticed that the results obtained differed. The specific model was Mistral-7B base and it was loaded in half-precision (`bfloat16`). While the DeepSpeed (blue) loss had converged well, the FSDP (orange) loss was not decreasing, as can be seen in Figure 1.
```

> ![Figure 1](https://cdn-lfs.huggingface.co/datasets/huggingface/documentation-images/1666e2fac64a29bdb6993c3856cc2964db208be8640f5d9e23b4a2111c0c800c?response-content-disposition=inline%3B+filename*%3DUTF-8%27%27figure_1.png%3B+filename%3D%22figure_1.png%22%3B&response-content-type=image%2Fpng&Expires=1718385931&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTcxODM4NTkzMX19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5odWdnaW5nZmFjZS5jby9kYXRhc2V0cy9odWdnaW5nZmFjZS9kb2N1bWVudGF0aW9uLWltYWdlcy8xNjY2ZTJmYWM2NGEyOWJkYjY5OTNjMzg1NmNjMjk2NGRiMjA4YmU4NjQwZjVkOWUyM2I0YTIxMTFjMGM4MDBjP3Jlc3BvbnNlLWNvbnRlbnQtZGlzcG9zaXRpb249KiZyZXNwb25zZS1jb250ZW50LXR5cGU9KiJ9XX0_&Signature=hNbNoQ8MClNsqOVTlWLAZF-3MJ5xIYoue7BoImRSoZyV3tb0lqYYqssLsLRCFE95PSkMj%7Eq0Yl7CGAEwAdTovDgeze5G46vNCDdkS8f5IXNHk%7ExCKQ4M0nRV4RQK9tDYiQOxFjRnU8STKtZtV8SejnVC1EapBgiNyCuptOWfpKAmsPqVds4LoiVMfHfk4ZV1Q41O4HIsLl4F9Iwnl37kHeykNb0tZgaKU8JaIoGBuDmpRESc3pjuQ2OHZN4TktQf4yaXuprAN2iuojVPHyc8KRMf3JOldq9yL22dynIvg6EWsx8iZmKaNHQuCu7OEa-694iYl61wqaQNiexE9qkcOw__&Key-Pair-Id=KVTP0A1DKRTAX)

Reviewer 2: I don't think you need all this path/cdn link (check the other articles) - the resolve path is fine

![Figure 3](https://cdn-lfs.huggingface.co/datasets/huggingface/documentation-images/e86c69648c3a6bfc8709c5149e28b29dbd5d61af621201175e97abbcf57a26ee?response-content-disposition=inline%3B+filename*%3DUTF-8%27%27figure_3.png%3B+filename%3D%22figure_3.png%22%3B&response-content-type=image%2Fpng&Expires=1718385978&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTcxODM4NTk3OH19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5odWdnaW5nZmFjZS5jby9kYXRhc2V0cy9odWdnaW5nZmFjZS9kb2N1bWVudGF0aW9uLWltYWdlcy9lODZjNjk2NDhjM2E2YmZjODcwOWM1MTQ5ZTI4YjI5ZGJkNWQ2MWFmNjIxMjAxMTc1ZTk3YWJiY2Y1N2EyNmVlP3Jlc3BvbnNlLWNvbnRlbnQtZGlzcG9zaXRpb249KiZyZXNwb25zZS1jb250ZW50LXR5cGU9KiJ9XX0_&Signature=r4pufHy2LN0gemtuivnUDVyaDiNkssKsfD-6K%7EGfao%7EoRwOS9IRe4AqiIZkyFIpnbs4yRZCrTdhIlfzxZyQzlnkia29CuEGIujiHo4uR3wTssI06GutEvaxDGzPnfkOiNqogr24BySHzEq1cBKytaiIzqlTbETcRkeI-FckCZ9a3wjnEEp%7EPd1oy1HYhAARlbPWJrzhBtuNgMsHjG7bA0WqfiuX-BuwYNVXuyfH2uWV4SKZ3pqz-P%7EAmmOSmvat3Yu2PZmtbe6grJGymMSqCZy%7Ej-RzwmrNFLNf14M9WA-P7MgYoB1dHDKosdj7CA9V67eevYIV3RiQSQvF37SYj9A__&Key-Pair-Id=KVTP0A1DKRTAX)

## Precision Matters

> Inside the `DeepSpeed` codebase, specifically, in the implementation of

Reviewer 2: ```suggestion
Inside the `DeepSpeed` codebase, specifically in the implementation of
```

| Training Step (forward, backward, reduction)                                             | ❌         | FSDP<br>DeepSpeed | follows [fsdp.MixedPrecision](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.MixedPrecision)<br>follows `deepspeed_config_file` mixed precision settings |
| Optimizer (pre-step)                                                                     | ✅         | FSDP<br>DeepSpeed | upcasting (if any) to `torch_dtype`<br>upcasting everything to `float32`                                                                                                   |
| Optimizer (actual step)                                                                  | ✅         | FSDP<br>DeepSpeed | occurs in `torch_dtype`<br>occurs in `float32`                                                                                                                             |

> Table 1: Summary of how FSDP and DeepSpeed handle mixed precision

A few takeaway points:

> * As noted an [🤗 Accelerate issue](https://github.com/huggingface/accelerate/issues/2624#issuecomment-2058402753), a rule of thumb when performing mixed precision is to keep trainable parameters in torch.float32. 

Reviewer 2: For consistency
```suggestion
* As noted an [🤗 Accelerate issue](https://github.com/huggingface/accelerate/issues/2624#issuecomment-2058402753), a rule of thumb when performing mixed precision is to keep trainable parameters in `float32`. 
```

> * Upcasting, as is done in `DeepSpeed`, may have negligible effect on memory consumption when sharding over a large number of GPUs. However, when using `DeepSpeed` on a small number of GPUs, the 2x increase in memory consumption can be significant.

Reviewer 2: ```suggestion
* Upcasting, as is done in `DeepSpeed`, may have a negligible effect on memory consumption when sharding over a large number of GPUs. However, when using `DeepSpeed` on a small number of GPUs, the 2x increase in memory consumption can be significant.
```

The result of this PR is to allow FSDP to operate in two modes:
- A “mixed-precision” mode like the DeepSpeed counterpart
- A low precision mode for memory constrained scenarios, as shown in Figure 4.

The two new FSDP modes are summarized in Table 2 and compared with DeepSpeed.
| **Framework**             | **Model Loading (`torch_dtype`)** | **Mixed Precision** | **Preparation (Local)** | **Training** | **Optimizer (Local)** |
| ------------------------- | --------------------------------- | ------------------- | ----------------------- | ------------ | --------------------- |
| FSDP (memory-constrained) | `bf16`                            | default (none)      | `bf16`                  | `bf16`       | `bf16`                |

> | FSDP (aligned mode)       | `bf16`                            | `bf16`              | `fp32`                  | `bf16`       | `fp32`                |

Reviewer 3: ```suggestion
| FSDP (mixed precision mode)       | `bf16`                            | `bf16`              | `fp32`                  | `bf16`       | `fp32`                |
```

* How is weight prefetching managed in FSDP and DeepSpeed?
* What is the equivalent of FSDP wrapping in DeepSpeed?

We consider various modes of configuring these frameworks in 🤗 Accelerate,
- From the command line during `accelerate launch`
- From the various `Plugin` classes 🤗 Accelerate provides for (`DeepSpeed`)[https://huggingface.co/docs/accelerate/main/en/package_reference/deepspeed] and (`FSDP`)[https://huggingface.co/docs/accelerate/main/en/package_reference/fsdp]

🤗 Accelerate makes it almost **trivial** to switch between FSDP and DeepSpeed, with the majority of it being an Accelerate config file change (see the new concept guide for instructions on this).

Besides the config
change, some of the other considerations (also outlined in the guide) are differences in how checkpoints are handled, etc.

All experiments in this blog can be reproduced with the code from the [original 🤗 Accelerate issue](https://github.com/huggingface/accelerate/issues/2624).

We intend to follow up with throughput comparisons at scale and techniques to better utilize those GPUs for tuning and alignment jobs while maintaining model quality.
## Acknowledgements

> This is an effort that involved several teams across multiple organizations to come together. It started at IBM Research, specifically Aldo Pareja who found the issue and Fabian Lim who identified the precision gaps and fixed this issue. Zach Mueller and [Stas Bekman](https://github.com/stas00) have been phenomenal in providing feedback and the fixes to accelerate. Less Wright from the PyTorch Team at Meta was very helpful in questions on FSDP parameters. Finally, we would also like to thank the [DeepSpeed](https://www.deepspeed.ai/) team for providing feedback on this blog.

Reviewer 2: ```suggestion
This is an effort that involved several teams across multiple organizations to come together. It started at IBM Research, specifically Aldo Pareja, who found the issue, and Fabian Lim, who identified the precision gaps and fixed this issue. Zach Mueller and [Stas Bekman](https://github.com/stas00) have been phenomenal in providing feedback and the fixes to accelerate. Less Wright from the PyTorch Team at Meta was very helpful with questions on FSDP parameters. Finally, we would also like to thank the [DeepSpeed](https://www.deepspeed.ai/) team for providing feedback on this blog.
```

Reviewer 3: Thanks @osanseviero for your suggestions and edits! :)

---
title: "From DeepSpeed to FSDP and Back Again with Hugging Face Accelerate"
authors:

> - user: mirinflim

Reviewer 3: ```suggestion
- user: mirinflim
- user: aldopareja
```

---

> title: "From DeepSpeed to FSDP and Back Again"

Reviewer 3: ```suggestion
title: "From DeepSpeed to FSDP and Back Again"
```
title: "A Hugging Face Accelerate Story of Multiple Backends"

Author: Went somewhere in the middle:

> From DeepSpeed to FSDP and Back Again with Hugging Face Accelerate

authors:
- user: mirinflim

> - user: rganti

Reviewer 3: ```suggestion
- user: rganti
```

Reviewer 3: Removing my name, keeping it to the folks who actually did the coding work :)

- user: muellerzr
- user: stas00
---

> # From DeepSpeed to FSDP and Back Again: How Interchangeable are the Frameworks?

Reviewer 3: ```suggestion
# From DeepSpeed to FSDP and Back Again: How Interchangeable are the Frameworks?
```
# A Hugging Face Accelerate Story of Multiple Backends: FSDP and DeepSpeed

Author: See prior note, adapted to:
> From DeepSpeed to FSDP and Back Again With Hugging Face Accelerate: How Interchangeable Are They?

There are two popular implementations of the [Zero Redundancy Optimizer (ZeRO)](https://arxiv.org/abs/1910.02054) algorithm in the community, one
from [Microsoft (DeepSpeed)](https://github.com/microsoft/DeepSpeed) and the other from [PyTorch Fully Sharded Data Parallelism, (FSDP)](https://pytorch.org/docs/stable/fsdp.html). [Hugging Face Accelerate](https://github.com/huggingface/accelerate) exposes both
frameworks for the end users to train/tune their models, and as a result is
upstreamed to the majority of the Hugging Face stack.

This blog highlights some fundamental differences between these frameworks.

> ## Are FSDP and DeepSpeed Interchangeable?

Reviewer 3: ```suggestion
## Are FSDP and DeepSpeed Interchangeable?
```
Are FSDP and DeepSpeed Interchangeable in HF Accelerate?

Author: I feel like the vibe of the article is using Accelerate as a medium to do this interchanging, moreso than it focused on Accelerate specifically, so I think we should leave it fairly broad there on purpose since it applies to more than just accelerate

Reviewer 3: Isn't it more of how HF accelerate exposed the question than DS or FSDP as frameworks? The upcasting is happening in accelerate DS integration or outside?

Author: It's happening inside, but I view accelerate as the medium towards comparing FSDP and DS 1:1, as you need to do things in FSDP to get that one to one, aka upcasting.

Reviewer 3: right, which is why I think the blog is how these frameworks are integrated rather than a broad comparison.

Author: After reading the blog though, it still again uses accelerate as a medium for doing this change, but the focus of the whole ordeal was making FSDP and DeepSpeed alike, so I think it still stands that it should be general.

Reviewer 3: ok, back and forth and we keep title the same

---
title: "From DeepSpeed to FSDP and Back Again"
authors:
- user: mirinflim
- user: rganti
- user: muellerzr
- user: stas00
---

# From DeepSpeed to FSDP and Back Again: How Interchangeable are the Frameworks?

> There are two popular implementations of the [ZeRO](https://arxiv.org/abs/1910.02054) algorithm in the community, one

Reviewer: ```suggestion
There are two popular implementations of the [Zero Redundancy Optimizer (ZeRO)](https://arxiv.org/abs/1910.02054) algorithm in the community, one
```

> from [Microsoft (DeepSpeed)](https://github.com/microsoft/DeepSpeed) and the other from [PyTorch (Fully Sharded Data Parallelism, or FSDP as it will be referred to throughout the rest of this blog)](https://pytorch.org/docs/stable/fsdp.html). [Hugging Face Accelerate](https://github.com/huggingface/accelerate) exposes both

Reviewer: ```suggestion
from [Microsoft (DeepSpeed)](https://github.com/microsoft/DeepSpeed) and the other from [PyTorch Fully Sharded Data Parallelism, (FSDP)](https://pytorch.org/docs/stable/fsdp.html). [Hugging Face Accelerate](https://github.com/huggingface/accelerate) exposes both
```

> these frameworks for the end users to train/tune their models, and as a result is

Reviewer: ```suggestion
frameworks for the end users to train/tune their models, and as a result is
```

> upstreamed to the majority of the Hugging Face stack. This blog highlights some
> fundamental differences between these frameworks, and has been summarized into an official [concept
> guide](https://huggingface.co/docs/accelerate/v0.31.0/concept_guides/fsdp_and_deepspeed) so that end users can switch seamlessly between these frameworks.

Reviewer: Already linked at the end so I would remove it here so that it doesn't interrupt the flow :)

```suggestion
upstreamed to the majority of the Hugging Face stack.

This blog highlights some fundamental differences between these frameworks.
```

## Are FSDP and DeepSpeed Interchangeable?

Recently we ported a training pipeline from DeepSpeed to FSDP. We were surprised when
the results obtained differed *substantially*. The specific model used was a Mistral-7B loaded in
half-precision (`bfloat16`). While the DeepSpeed (blue) loss had converged well, the FSDP (orange)
loss was not decreasing, as can be seen in Figure 1.

[[ZACH: INSERT FIGURE ONE HERE]]

Initially we thought that FSDP may be computing the mean of the gradients rather than the sum,
hence requiring a higher learning rate. So, we bumped up the learning rate by 4x since we were
using 4 GPUs. Then we saw the following loss behavior, shown in Figure 2:

[[ZACH: INSERT FIGURE TWO HERE]]

It looked as though scaling the learning rate had indeed helped improve the reproducibility between the two frameworks! However, our next experiment debunked this hypothesis. 

> After setting both FSDP and DeepSpeed to a higher learning rate (`1e-5` from `1e-6` initially), and we observed similar loss and gradient norm characteristics, shown in Figure 3:

Reviewer: ```suggestion
After setting both FSDP and DeepSpeed to a higher learning rate (`1e-5` from `1e-6` initially), we observed similar loss and gradient norm characteristics, shown in Figure 3:
```

## The Investigation Unfolds: Precision Matters

Inside the `DeepSpeed` codebase, specifically, in the implementation of

> `DeepSpeedZeroOptimizer_Stage3` (as the name implies, what handles doing Stage 3 optimizer sharding), we noticed that the `trainable_param_groups` (the parameter groups being trained on) pass through an 

Reviewer: ```suggestion
`DeepSpeedZeroOptimizer_Stage3` (as the name implies, what handles doing Stage 3 optimizer sharding), we noticed that the `trainable_param_groups`, the parameter groups being trained on, pass through an 
```

internal `_setup_for_real_optimizer` function call, which calls another function called `_create_fp32_partitions`.

> Essentially `DeepSpeed` was performing upcasting from `bf16` to `fp32` internally, and it always kept the master weights in 
> `fp32` (full precision) *by design*. This upcasting meant that the optimizer could converge at learning rates that struggled
> in lower precisions. In Figure 1, the FSDP was not coverging with `1e-6`, and was just a coincidence that bumping up the learning rate to
> `4e-6` (4x) just happened to be sufficient for converging, as show in in Figure 2. 

Reviewer: Referencing the function (`create_fp32_partitions`) instead of DeepSpeed will help users better connect these two paragraphs.

```suggestion
This function was upcasting from `bf16` to `fp32` (full precision) internally, and it always kept the master weights in  `fp32` *by design*. This upcasting meant that the optimizer could converge at learning rates that struggled in lower precisions. In Figure 1, FSDP was not converging with `1e-6`, and it was just a coincidence that bumping up the learning rate to `4e-6` (4x) just happened to be sufficient for converging, as shown in Figure 2. 
```

The crux of the issue is how these two frameworks handle training on and utilizing lower precisions.

During distributed training, before the model and optimizer parameters are split across GPUs they are first "flattened" to
a one-dimensional `torch.Tensor` (or, a list). FSDP and DeepSpeed use different `dtype`s for these "flattened"
parameters which has ramifications for PyTorch Optimizers. Table 1 outlines the processes for both

> frameworks; the "Local" column indicates if the process occurring per-GPU, therefore the memory
> overhead from upcasting is amortized by the number of GPUs.

Reviewer: I would pick another word instead of "amortized" like maybe "reduced". It gives "mortgage" and "loan" vibes

```suggestion
frameworks. The "Local" column indicates if the process occurs per-GPU, and therefore the memory
overhead from upcasting is amortized by the number of GPUs.
```

| Training Step (forward, backward, reduction)                                             | ❌         | FSDP<br>DeepSpeed | follows [fsdp.MixedPrecision](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.MixedPrecision)<br>follows `deepspeed_config_file` mixed precision settings |
| Optimizer (pre-step)                                                                     | ✅         | FSDP<br>DeepSpeed | upcasting (if any) to `torch_dtype`<br>upcasting everything to `float32`                                                                                                   |
| Optimizer (actual step)                                                                  | ✅         | FSDP<br>DeepSpeed | occurs in `torch_dtype`<br>occurs in `float32`                                                                                                                             |

> Table 1: Summary of how FSDP and DeepSpeed handle mixed precision

> A few takeaway points:

Reviewer: ```suggestion
A few takeaways:
```

* As noted in the [🤗 Accelerate Documentation], a rule of thumb when performing mixed precision
is to keep trainable parameters in `torch.float32`.
* The upcasting done by `DeepSpeed` may have negligible effect on memory
consumption when sharding over a large number of GPUs, however when using
a small number of GPUs the 2x increase in memory
consumption can be **significant**.
* The torch-native implementation of FSDP does not force upcasting, allowing a
user to operate `torch.Optimizer` in low precision. This offers more flexibility than
the native upcasting of `DeepSpeed`.

> We created a 🤗 Accelerate pull request that was included in the 0.30.0 release that aimed to improve this.

Reviewer: I think it's ok to remove this sentence since it is mentioned again in the next paragraph.

```suggestion
```

| FSDP (aligned mode) | 3158.7                    | 10.4              | 0.41                              |
| DeepSpeed           | 3094.5                    | 10.6              | 0.40                              |

> Table 3: Ballpark throughput comparisons between FSDP and DeepSpeed on four A100 GPUs.

## Closing thoughts

We provided a new [concept guide](https://huggingface.co/docs/accelerate/v0.31.0/concept_guides/fsdp_and_deepspeed) to help users migrate between the two frameworks. The guide helps users answer questions such as:

* How do we achieve equivalent sharding strategies?
* How do we perform efficient model loading?
* How is weight prefetching managed in FSDP and DeepSpeed?
* What is the equivalent of FSDP wrapping in DeepSpeed?

We considered various modes of configuring these in 🤗 Accelerate as well:
- From the command line during `accelerate launch`
- From the various `Plugin` classes 🤗 Accelerate provides for (`DeepSpeed`)[https://huggingface.co/docs/accelerate/main/en/package_reference/deepspeed] and (`FSDP`)[https://huggingface.co/docs/accelerate/main/en/package_reference/fsdp]

> 🤗 accelerate makes it almost **trivial** to switch between FSDP and DeepSpeed, with the majority of it being an
> accelerate config file change, see the new concept guide for instructions on this. Besides the config

Reviewer: ```suggestion
🤗 Accelerate makes it almost **trivial** to switch between FSDP and DeepSpeed, with the majority of it being an
Accelerate config file change (see the new concept guide for instructions on this). Besides the config
```

---
title: "From DeepSpeed to FSDP and Back Again with Hugging Face Accelerate"
authors:
- user: mirinflim
- user: muellerzr
- user: stas00
---

> # From DeepSpeed to FSDP and Back Again With Hugging Face Accelerate: How Interchangeable Are They?

Reviewer 3: ```suggestion
# From DeepSpeed to FSDP and Back Again With Hugging Face Accelerate: How Interchangeable Are They?
```
A Hugging Face Accelerate Story of Multiple Backends: FSDP and DeepSpeed

> There are two popular implementations of the [Zero Redundancy Optimizer (ZeRO)](https://arxiv.org/abs/1910.02054) algorithm in the community, one
> from [Microsoft (DeepSpeed)](https://github.com/microsoft/DeepSpeed) and the other from [PyTorch Fully Sharded Data Parallelism, (FSDP)](https://pytorch.org/docs/stable/fsdp.html). [Hugging Face Accelerate](https://github.com/huggingface/accelerate) exposes both
> frameworks for the end users to train/tune their models, and as a result is
> upstreamed to the majority of the Hugging Face stack.
> 
> This blog highlights some fundamental differences between these frameworks.

Reviewer 3: ```suggestion
There are two popular implementations of the [Zero Redundancy Optimizer (ZeRO)](https://arxiv.org/abs/1910.02054) algorithm in the community, one
from [Microsoft (DeepSpeed)](https://github.com/microsoft/DeepSpeed) and the other from [PyTorch Fully Sharded Data Parallelism, (FSDP)](https://pytorch.org/docs/stable/fsdp.html). [Hugging Face Accelerate](https://github.com/huggingface/accelerate) exposes both
frameworks for the end users to train/tune their models, and as a result is
upstreamed to the majority of the Hugging Face stack.

This blog highlights some fundamental differences between these frameworks.
```
There are two popular implementations of the [ZeRO](https://arxiv.org/abs/1910.02054) algorithm in the community, one from [DeepSpeed](https://github.com/microsoft/DeepSpeed) and the other from [PyTorch](https://pytorch.org/docs/stable/fsdp.html). HuggingFace [Accelerate](https://huggingface.co/docs/accelerate/en/index) exposes both these frameworks for the end users to train/tune their models. This blog highlights the differences between how these backends are exposed through Accelerate. To enable users to seamlessly switch between these backends, we [upstreamed a precision related change](https://github.com/huggingface/accelerate/issues/2624) and a [concept guide](https://huggingface.co/docs/accelerate/concept_guides/fsdp_and_deepspeed).

## Are FSDP and DeepSpeed Interchangeable?

> Recently we ported a training pipeline from DeepSpeed to FSDP. We were surprised when
> the results obtained differed *substantially*. The specific model used was a Mistral-7B loaded in
> half-precision (`bfloat16`). While the DeepSpeed (blue) loss had converged well, the FSDP (orange)
> loss was not decreasing, as can be seen in Figure 1.
> 
> Initially we thought that FSDP may be computing the mean of the gradients rather than the sum,
> hence requiring a higher learning rate. So, we bumped up the learning rate by 4x since we were
> using 4 GPUs. Then we saw the following loss behavior, shown in Figure 2:

Reviewer 3: ```suggestion
Recently we ported a training pipeline from DeepSpeed to FSDP. We were surprised when
the results obtained differed *substantially*. The specific model used was a Mistral-7B loaded in
half-precision (`bfloat16`). While the DeepSpeed (blue) loss had converged well, the FSDP (orange)
loss was not decreasing, as can be seen in Figure 1.

Initially we thought that FSDP may be computing the mean of the gradients rather than the sum,
hence requiring a higher learning rate. So, we bumped up the learning rate by 4x since we were
using 4 GPUs. Then we saw the following loss behavior, shown in Figure 2:
```

Recently we tried running a training pipeline with DeepSpeed and PyTorch FSDP (at the time of the port, we used Accelerate v0.28 that used DeepSpeed v0.14 and Torch 2.2.1). We noticed that the results obtained differed. The specific model was Mistral-7B base and it was loaded in half-precision: `bfloat16`. While the DeepSpeed (blue) loss had converged well, the FSDP (orange) loss was not decreasing, as can be seen in Figure 1.

Author: As we said before, unless we know the full exact versions we are not including the versions. I personally prefer the other version, but we can include this without the versions instead.

Reviewer 3: removed the versions in the suggestion

[[ZACH: INSERT FIGURE TWO HERE]]

> It looked as though scaling the learning rate had indeed helped improve the reproducibility between the two frameworks! However, our next experiment debunked this hypothesis. 
> 
> After setting both FSDP and DeepSpeed to a higher learning rate (`1e-5` from `1e-6` initially), we observed similar loss and gradient norm characteristics, shown in Figure 3:
> 

Reviewer 3: ```suggestion
It looked as though scaling the learning rate had indeed helped improve the reproducibility between the two frameworks! However, our next experiment debunked this hypothesis. 

After setting both FSDP and DeepSpeed to a higher learning rate (`1e-5` from `1e-6` initially), we observed similar loss and gradient norm characteristics, shown in Figure 3:
```
We hypothesized that the learning rate may need scaling by number of GPUs and bumped up the learning rate by 4x since we were using 4 GPUs. Then, we saw the following loss behavior, shown in Figure 2. 

It looked like the desired behavior had been achieved by scaling the FSDP learning rate by the number of GPUs! However,  when we tried a different learning rate (1e-5) without scaling the learning rate, we observed similar loss and gradient norm characteristics for both frameworks, shown in Figure 3.

## The Investigation Unfolds: Precision Matters

Inside the `DeepSpeed` codebase, specifically, in the implementation of
`DeepSpeedZeroOptimizer_Stage3` (as the name implies, what handles doing Stage 3 optimizer sharding), we noticed that the `trainable_param_groups`, the parameter groups being trained on, pass through an 
internal `_setup_for_real_optimizer` function call, which calls another function called `_create_fp32_partitions`.

> This function was upcasting from `bf16` to `fp32` (full precision) internally, and it always kept the master weights in  `fp32` *by design*. This upcasting meant that the optimizer could converge at learning rates that struggled in lower precisions. In Figure 1, FSDP was not converging with `1e-6`, and it was just a coincidence that bumping up the learning rate to `4e-6` (4x) just happened to be sufficient for converging, as shown in Figure 2. 

Reviewer: ```suggestion
This function was upcasting from `bf16` to `fp32` (full precision) internally, and it always kept the master weights in `fp32` *by design*. This upcasting meant that the optimizer could converge at learning rates that struggled in lower precisions. In Figure 1, FSDP was not converging with `1e-6`, and it was just a coincidence that bumping up the learning rate to `4e-6` (4x) just happened to be sufficient for converging, as shown in Figure 2. 
```

> ## The Investigation Unfolds: Precision Matters
> 
> Inside the `DeepSpeed` codebase, specifically, in the implementation of
> `DeepSpeedZeroOptimizer_Stage3` (as the name implies, what handles doing Stage 3 optimizer sharding), we noticed that the `trainable_param_groups`, the parameter groups being trained on, pass through an 
> internal `_setup_for_real_optimizer` function call, which calls another function called `_create_fp32_partitions`.
> 
> This function was upcasting from `bf16` to `fp32` (full precision) internally, and it always kept the master weights in `fp32` *by design*. This upcasting meant that the optimizer could converge at learning rates that struggled in lower precisions. In Figure 1, FSDP was not converging with `1e-6`, and it was just a coincidence that bumping up the learning rate to `4e-6` (4x) just happened to be sufficient for converging, as shown in Figure 2. 
> 
> The crux of the issue is how these two frameworks handle training on and utilizing lower precisions.
> 
> During distributed training, before the model and optimizer parameters are split across GPUs they are first "flattened" to
> a one-dimensional `torch.Tensor` (or, a list). FSDP and DeepSpeed use different `dtype`s for these "flattened"
> parameters which has ramifications for PyTorch Optimizers. Table 1 outlines the processes for both
> frameworks. The "Local" column indicates if the process occurs per-GPU, and therefore the memory
> overhead from upcasting is amortized by the number of GPUs.

Reviewer 3: ```suggestion
## The Investigation Unfolds: Precision Matters

Inside the `DeepSpeed` codebase, specifically, in the implementation of
`DeepSpeedZeroOptimizer_Stage3` (as the name implies, what handles doing Stage 3 optimizer sharding), we noticed that the `trainable_param_groups`, the parameter groups being trained on, pass through an 
internal `_setup_for_real_optimizer` function call, which calls another function called `_create_fp32_partitions`.

This function was upcasting from `bf16` to `fp32` (full precision) internally, and it always kept the master weights in `fp32` *by design*. This upcasting meant that the optimizer could converge at learning rates that struggled in lower precisions. In Figure 1, FSDP was not converging with `1e-6`, and it was just a coincidence that bumping up the learning rate to `4e-6` (4x) just happened to be sufficient for converging, as shown in Figure 2. 

The crux of the issue is how these two frameworks handle training on and utilizing lower precisions.

During distributed training, before the model and optimizer parameters are split across GPUs they are first "flattened" to
a one-dimensional `torch.Tensor` (or, a list). FSDP and DeepSpeed use different `dtype`s for these "flattened"
parameters which has ramifications for PyTorch Optimizers. Table 1 outlines the processes for both
frameworks. The "Local" column indicates if the process occurs per-GPU, and therefore the memory
overhead from upcasting is amortized by the number of GPUs.
```
We notice that inside the DeepSpeed codebase, specifically      in the implementation of `DeepSpeedZeroOptimizer_Stage3`, the `trainable_param_groups` pass through an internal `_setup_for_real_optimizer` function call, which calls another function called `_create_fp32_partitions`.
As the `fp32` term in the name suggests, DeepSpeed was performing upcasting internally, and it always keeps its master weights in `fp32` by design. This upcasting to full precision meant that the optimizer could converge at learning rates that it would not converge in lower precision. The earlier observations were artifacts of this precision difference.

In FSDP, before the model and optimizer parameters are distributed across GPUs, they are first "flattened" to a one-dimensional [torch.Tensor](https://pytorch.org/docs/stable/tensors.html#torch-tensor). FSDP and DeepSpeed use different `dtype` for these "flattened" parameters which has ramifications for [torch.Optimizer](https://pytorch.org/docs/stable/optim.html#module-torch.optim). Table 1 outlines the processes for both frameworks; the "Local" column indicates the process occurring per-GPU, therefore the memory overhead from upcasting is amortized by the number of GPUs.

| Training Step (forward, backward, reduction)                                             | ❌         | FSDP<br>DeepSpeed | follows [fsdp.MixedPrecision](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.MixedPrecision)<br>follows `deepspeed_config_file` mixed precision settings |
| Optimizer (pre-step)                                                                     | ✅         | FSDP<br>DeepSpeed | upcasting (if any) to `torch_dtype`<br>upcasting everything to `float32`                                                                                                   |
| Optimizer (actual step)                                                                  | ✅         | FSDP<br>DeepSpeed | occurs in `torch_dtype`<br>occurs in `float32`                                                                                                                             |

> Table 1: Summary of how FSDP and DeepSpeed handle mixed precision

> A few takeaways:
> * As noted in the [🤗 Accelerate Documentation], a rule of thumb when performing mixed precision
> is to keep trainable parameters in `torch.float32`.
> * The upcasting done by `DeepSpeed` may have negligible effect on memory
> consumption when sharding over a large number of GPUs, however when using
> a small number of GPUs the 2x increase in memory
> consumption can be **significant**.
> * The torch-native implementation of FSDP does not force upcasting, allowing a
> user to operate `torch.Optimizer` in low precision. This offers more flexibility than
> the native upcasting of `DeepSpeed`.

Reviewer 3: ```suggestion
A few takeaways:
* As noted in the [🤗 Accelerate Documentation], a rule of thumb when performing mixed precision
is to keep trainable parameters in `torch.float32`.
* The upcasting done by `DeepSpeed` may have negligible effect on memory
consumption when sharding over a large number of GPUs, however when using
a small number of GPUs the 2x increase in memory
consumption can be **significant**.
* The torch-native implementation of FSDP does not force upcasting, allowing a
user to operate `torch.Optimizer` in low precision. This offers more flexibility than
the native upcasting of `DeepSpeed`.
```
A few takeaway points:
* As noted in the 🤗  Documentation, a rule of thumb when performing mixed precision is to keep trainable parameters in torch.float32. 
* Upcasting, as is done in DeepSpeed, may have negligible effect on memory consumption when sharding over a large number of GPUs. However, when using DeepSpeed on a small number of GPUs, the 2x increase in memory consumption can be significant.
* The torch-native implementation of FSDP does not force upcasting, allowing a user to operate torch.Optimizer in low precision. This offers more flexibility than the native upcasting of DeepSpeed.

Author: I've updated this a bit to read better. Let me know if you are okay with it

Reviewer 3: am good with it!

## Harmonizing DeepSpeed and FSDP in 🤗 Accelerate

To better align DeepSpeed and FSDP in 🤗 Accelerate, [Stas Bekman](https://github.com/stas00) suggested performing the upcasting automatically
for FSDP when mixed precision was enabled. This was completed in [this accelerate PR](https://github.com/huggingface/accelerate/pull/2674) and has been
available since the 0.30.0 release.

> 

Author: Let me know if you are okay with the edits here

[[ZACH: INSERT FIGURE FOUR HERE]]

> The result of this allowed FSDP to operate in two modes:
> - A "mixed-precision" mode which uses a similar behavior to DeepSpeed
> - A low precision mode for memory-constrained scenarios, as shown in Figure 4.
> 
> The two new FSDP modes are summarized below and compared to DeepSpeed, with the second mode achieving complete parity with DeepSpeed:
> 

Reviewer 3: ```suggestion
The result of this allowed FSDP to operate in two modes:
- A "mixed-precision" mode which uses a similar behavior to DeepSpeed
- A low precision mode for memory-constrained scenarios, as shown in Figure 4.

The two new FSDP modes are summarized below and compared to DeepSpeed, with the second mode achieving complete parity with DeepSpeed:

```
The result of this PR is to allow FSDP to operate in two modes, a “mixed-precision” mode like the DeepSpeed counterpart, and a low precision mode for memory constrained scenarios, as shown in Figure 5.
The two new FSDP modes are summarized in Figure 6 and compared with DeepSpeed.

| **Framework**             | **Model Loading (`torch_dtype`)** | **Mixed Precision** | **Preparation (Local)** | **Training** | **Optimizer (Local)** |
| ------------------------- | --------------------------------- | ------------------- | ----------------------- | ------------ | --------------------- |
| FSDP (memory-constrained) | `bf16`                            | default (none)      | `bf16`                  | `bf16`       | `bf16`                |
| FSDP (aligned mode)       | `bf16`                            | `bf16`              | `fp32`                  | `bf16`       | `fp32`                |
| DeepSpeed                 | `bf16`                            | `bf16`              | `fp32`                  | `bf16`       | `fp32`                |

> Table 2: Summary of the two new FSDP modes and comparisons with DeepSpeed

> ## A Note on Throughput

Reviewer: ```suggestion
## Throughput results
```

> We use the [IBM `granite7b`](https://huggingface.co/ibm-granite/granite-7b-base) model (which follows the Meta Llama2 architecture) for throughput comparisons. We compare MFU and tokens/sec/GPU metrics and show them for FSDP (full sharding) and DeepSpeed (Zero3). 

Reviewer: ```suggestion
We use the [IBM `granite7b`](https://huggingface.co/ibm-granite/granite-7b-base) model (which follows the Meta Llama2 architecture) for throughput comparisons. We compare model flops utilization (MFU) and tokens/sec/GPU metrics and show them for FSDP (full sharding) and DeepSpeed (Zero3). 
```

> We use the [IBM `granite7b`](https://huggingface.co/ibm-granite/granite-7b-base) model (which follows the Meta Llama2 architecture) for throughput comparisons. We compare model flops utilization (MFU) and tokens/sec/GPU metrics and show them for FSDP (full sharding) and DeepSpeed (Zero3). 
> 
> We utilized four A100 GPUs with the following hyperparameters:
> - Batch size of 8
> - Model loaded in `torch.bfloat16`
> - Mixed precision is the same dtype.
>   
> Table 3 shows that FSDP and DeepSpeed are expected to perform relatively similarly.
> 
> > We intend to follow up with a comprehensive throughput comparison and approaches to improve throughput (e.g., 4D masks with packing, torch.compile, selective activation checkpointing) as large scale alignment techniques like InstructLab and GLAN become popular.
> 

Reviewer 3: ```suggestion
We use the [IBM `granite7b`](https://huggingface.co/ibm-granite/granite-7b-base) model (which follows the Meta Llama2 architecture) for throughput comparisons. We compare model flops utilization (MFU) and tokens/sec/GPU metrics and show them for FSDP (full sharding) and DeepSpeed (Zero3). 

We utilized four A100 GPUs with the following hyperparameters:
- Batch size of 8
- Model loaded in `torch.bfloat16`
- Mixed precision is the same dtype.
  
Table 3 shows that FSDP and DeepSpeed are expected to perform relatively similarly.

> We intend to follow up with a comprehensive throughput comparison and approaches to improve throughput (e.g., 4D masks with packing, torch.compile, selective activation checkpointing) as large scale alignment techniques like InstructLab and GLAN become popular.

```
We use the [IBM Granite 7B](https://huggingface.co/ibm-granite/granite-7b-base) model (which follows the Meta Llama2 architecture) for throughput comparisons. We compare MFU and tokens/sec/GPU metrics and show them for FSDP (full sharding) and DeepSpeed (Zero3). We used four A100 GPUs as before with a batch size of 8, loaded model in torch.bfloat16 and mixed precision is the same dtype. Figure 7 shows that FSDP and DeepSpeed are expected to perform similarly. We intend to follow up with a comprehensive throughput comparison and approaches to improve throughput (e.g., 4D masks with packing, torch.compile, selective activation checkpointing) as large scale alignment techniques like [InstructLab](https://github.com/instructlab) and [GLAN](https://arxiv.org/abs/2402.13064) become popular.

| **Framework**       | **Tokens / sec / device** | **Step time (s)** | **Model Flops Utilization (MFU)** |
| ------------------- | ------------------------- | ----------------- | --------------------------------- |
| FSDP (aligned mode) | 3158.7                    | 10.4              | 0.41                              |
| DeepSpeed           | 3094.5                    | 10.6              | 0.40                              |

> Table 3: Ballpark throughput comparisons between FSDP and DeepSpeed on four A100 GPUs.

## Closing thoughts

> We provided a new [concept guide](https://huggingface.co/docs/accelerate/v0.31.0/concept_guides/fsdp_and_deepspeed) to help users migrate between the two frameworks. The guide helps users answer questions such as:
> 
> * How do we achieve equivalent sharding strategies?
> * How do we perform efficient model loading?
> * How is weight prefetching managed in FSDP and DeepSpeed?
> * What is the equivalent of FSDP wrapping in DeepSpeed?
> 
> We considered various modes of configuring these in 🤗 Accelerate as well:
> - From the command line during `accelerate launch`
> - From the various `Plugin` classes 🤗 Accelerate provides for (`DeepSpeed`)[https://huggingface.co/docs/accelerate/main/en/package_reference/deepspeed] and (`FSDP`)[https://huggingface.co/docs/accelerate/main/en/package_reference/fsdp]
> 
> 🤗 Accelerate makes it almost **trivial** to switch between FSDP and DeepSpeed, with the majority of it being an
> Accelerate config file change (see the new concept guide for instructions on this). Besides the config
> change, some of the other considerations (also outlined in the guide) are differences in how checkpoints are handled, etc.
> 
> All experiments in this blog can be reproduced with the code from the [original 🤗 Accelerate issue](https://github.com/huggingface/accelerate/issues/2624).
> 
> We intend to follow up with throughput comparisons at scale and techniques to better utilize those GPUs for tuning and alignment jobs while maintaining model quality.
> 

Reviewer 3: ```suggestion
We provided a new [concept guide](https://huggingface.co/docs/accelerate/v0.31.0/concept_guides/fsdp_and_deepspeed) to help users migrate between the two frameworks. The guide helps users answer questions such as:

* How do we achieve equivalent sharding strategies?
* How do we perform efficient model loading?
* How is weight prefetching managed in FSDP and DeepSpeed?
* What is the equivalent of FSDP wrapping in DeepSpeed?

We considered various modes of configuring these in 🤗 Accelerate as well:
- From the command line during `accelerate launch`
- From the various `Plugin` classes 🤗 Accelerate provides for (`DeepSpeed`)[https://huggingface.co/docs/accelerate/main/en/package_reference/deepspeed] and (`FSDP`)[https://huggingface.co/docs/accelerate/main/en/package_reference/fsdp]

🤗 Accelerate makes it almost **trivial** to switch between FSDP and DeepSpeed, with the majority of it being an
Accelerate config file change (see the new concept guide for instructions on this). Besides the config
change, some of the other considerations (also outlined in the guide) are differences in how checkpoints are handled, etc.

All experiments in this blog can be reproduced with the code from the [original 🤗 Accelerate issue](https://github.com/huggingface/accelerate/issues/2624).

We intend to follow up with throughput comparisons at scale and techniques to better utilize those GPUs for tuning and alignment jobs while maintaining model quality.

```
As a note to the community, we created a [new concept guide](https://huggingface.co/docs/accelerate/main/en/concept_guides/fsdp_and_deepspeed) to help users migrate from DeepSpeed to FSDP and vice-versa and we will continue to maintain this guide with the community. The guide helps users answer questions like:
* How do we achieve equivalent sharding strategies?
* How do we perform efficient model loading?
* How is weight prefetching managed in FSDP and DeepSpeed?
* What is the equivalent of FSDP wrapping in DeepSpeed

We consider various modes of configuring these frameworks in 🤗 accelerate,
* Via accelerate launch
* Via the [FullyShardedDataParallelPlugin](https://huggingface.co/docs/accelerate/main/en/package_reference/fsdp#accelerate.FullyShardedDataParallelPlugin) and [DeepSpeedPlugin](https://huggingface.co/docs/accelerate/main/en/package_reference/deepspeed#accelerate.DeepSpeedPlugin),

🤗 accelerate makes it almost trivial to switch between FSDP and DeepSpeed, it is mostly an accelerate config file change; see the new concept guide for instructions on this. Besides the config change, some of the other considerations (also outlined in the guide) are the differences in the way  the checkpoints are handled, etc. All experiments in this blog can be reproduced with code from the [original HuggingFace issue](https://github.com/huggingface/accelerate/issues/2624).
We intend to follow up with throughput comparisons at scale and techniques to better utilize those GPUs for tuning and alignment jobs while maintaining model quality.

## Acknowledgements

> This is an effort that involved several teams across multiple organizations to come together. It started at IBM Research, specifically Aldo Pareja who found the issue and Fabian Lim who identified the precision gaps and fixed this issue. Zach Mueller and [Stas Bekman](https://github.com/stas00) have been phenomenal in providing feedback and the fixes to accelerate. Less Wright from the PyTorch Team at Meta was very helpful in questions on FSDP parameters. Finally, we would also like to thank the [DeepSpeed team](https://www.deepspeed.ai/) for providing feedback on this blog.

Reviewer 3: ```suggestion
This is an effort that involved several teams across multiple organizations to come together. It started at IBM Research, specifically Aldo Pareja who found the issue and Fabian Lim who identified the precision gaps and fixed this issue. Zach Mueller and [Stas Bekman](https://github.com/stas00) have been phenomenal in providing feedback and the fixes to accelerate. Less Wright from the PyTorch Team at Meta was very helpful in questions on FSDP parameters. Finally, we would also like to thank the [DeepSpeed team](https://www.deepspeed.ai/) for providing feedback on this blog.
```
This is an effort that involved several teams across multiple organizations to come together. It started at IBM Research, specifically Aldo Pareja who found the issue and Fabian Lim who identified the precision gaps and fixed this issue. Zach Mueller and [Stas Bekman](https://github.com/stas00) have been phenomenal in providing feedback and the fixes to accelerate. Less Wright from the PyTorch Team at Meta was very helpful in questions on FSDP parameters. Finally, we would also like to thank the [DeepSpeed](https://www.deepspeed.ai/) team for providing feedback on this blog.
